tutorial/python_agents/base_agent.py

```diff
@@ -0,0 +1,457 @@
+#!/usr/bin/env python3
+#
+#  Copyright (c) 2015 - 2022, Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+"""
+This package provides a BaseAgent class to monitor processes. Extend this class
+to implement additional agent functionality.
+
+The package provides parse_common_args and run_controller functions
+to create an argument parser for classes that extend BaseAgent.
+"""
+
+import sys
+import os
+import argparse
+from collections import OrderedDict
+import socket
+import yaml
+
+import geopmdpy
+import geopmdpy.runtime
+import geopmdpy.pio
+import geopmpy.reporter
+
+
+CSV_SEP_CHAR = "|"
+DOMAIN_SEP_CHAR = "-"
+TIME_SIGNAL = ("TIME", geopmdpy.topo.DOMAIN_BOARD, 0, "-")
+
+DEFAULT_SIGNAL_LIST = [
+    ("CPU_INSTRUCTIONS_RETIRED",           geopmdpy.topo.DOMAIN_BOARD, 0, "incr"),
+    ("CPU_POWER",                                       geopmdpy.topo.DOMAIN_BOARD, 0, "avg"), 
+    ("DRAM_POWER",                                   geopmdpy.topo.DOMAIN_BOARD, 0, "avg"),
+    ("CPU_FREQUENCY_STATUS",                 geopmdpy.topo.DOMAIN_BOARD, 0, "avg"), 
+    ("CPU_UNCORE_FREQUENCY_STATUS", geopmdpy.topo.DOMAIN_BOARD, 0, "avg"),  
+    ("CPU_CORE_TEMPERATURE",                geopmdpy.topo.DOMAIN_BOARD, 0, "avg"),
+    ("DRAM_ENERGY",                                 geopmdpy.topo.DOMAIN_BOARD, 0, "incr"),
+    ("CPU_ENERGY",                                     geopmdpy.topo.DOMAIN_BOARD, 0, "incr"), 
+]
+
+
+def generate_signal_list_from_user_list(user_list):
+    signal_list = DEFAULT_SIGNAL_LIST
+    if user_list is not None:
+        signal_specifications = user_list.split(',')
+        for specification in signal_specifications:
+            specification_parts = specification.split('@')
+            if len(specification_parts) == 1:
+                signal_name = specification_parts[0]
+                signal_domain = geopmdpy.topo.DOMAIN_BOARD
+            elif len(specification_parts) == 2:
+                signal_name, signal_domain = specification_parts
+            else:
+                print(f'Error: invalid signal specification: {specification}',
+                      file=sys.stderr)
+                exit(1)
+
+            _, _, behavior = geopmdpy.pio.signal_info(signal_name)
+            behavior_str = 'incr' if behavior == 1 else 'avg'
+
+            for domain_index in range(geopmdpy.topo.num_domain(signal_domain)):
+                signal_list.append((signal_name, signal_domain, domain_index, behavior_str))
+
+
+class BaseAgent(geopmdpy.runtime.Agent):
```

Comment:
Can we reduce any of the complexity here? Some of it was needed for now-obsolete use cases. e.g., try to do less in-agent accumulation